File: sdks/sandbox/javascript/src/manager.ts
// Copyright 2026 Alibaba Group Holding Ltd.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { ConnectionConfig, type ConnectionConfigOptions } from "./config/connection.js";
import { createDefaultAdapterFactory } from "./factory/defaultAdapterFactory.js";
import type { AdapterFactory } from "./factory/adapterFactory.js";

import type { ListSandboxesResponse, SandboxId, SandboxInfo } from "./models/sandboxes.js";
import type { Sandboxes } from "./services/sandboxes.js";

export interface SandboxManagerOptions {
  connectionConfig?: ConnectionConfig | ConnectionConfigOptions;
  adapterFactory?: AdapterFactory;
}

export interface SandboxFilter {
  states?: string[];
  metadata?: Record<string, string>;
  page?: number;
  pageSize?: number;
}

/**
 * Administrative interface for managing sandboxes (list/get/pause/resume/kill/renew).
 *
 * For interacting *inside* a sandbox, use {@link Sandbox}.
 */
export class SandboxManager {
  private readonly sandboxes: Sandboxes;
  private readonly connectionConfig: ConnectionConfig;

  private constructor(opts: { sandboxes: Sandboxes; connectionConfig: ConnectionConfig }) {
    this.sandboxes = opts.sandboxes;
    this.connectionConfig = opts.connectionConfig;
  }

  static create(opts: SandboxManagerOptions = {}): SandboxManager {
    const baseConnectionConfig = opts.connectionConfig instanceof ConnectionConfig
      ? opts.connectionConfig
      : new ConnectionConfig(opts.connectionConfig);
    const connectionConfig = baseConnectionConfig.withTransportIfMissing();
    const lifecycleBaseUrl = connectionConfig.getBaseUrl();
    const adapterFactory = opts.adapterFactory ?? createDefaultAdapterFactory();
    let sandboxes: Sandboxes;
    try {
      sandboxes = adapterFactory.createLifecycleStack({
        connectionConfig,
        lifecycleBaseUrl,
      }).sandboxes;
    } catch (err) {
      void connectionConfig.closeTransport().catch(() => undefined);
      throw err;
    }
    return new SandboxManager({ sandboxes, connectionConfig });
  }

  listSandboxInfos(filter: SandboxFilter = {}): Promise<ListSandboxesResponse> {
    return this.sandboxes.listSandboxes({
      states: filter.states,
      metadata: filter.metadata,
      page: filter.page,
      pageSize: filter.pageSize,
    });
  }

  getSandboxInfo(sandboxId: SandboxId): Promise<SandboxInfo> {
    return this.sandboxes.getSandbox(sandboxId);
  }

  killSandbox(sandboxId: SandboxId): Promise<void> {
    return this.sandboxes.deleteSandbox(sandboxId);
  }

  pauseSandbox(sandboxId: SandboxId): Promise<void> {
    return this.sandboxes.pauseSandbox(sandboxId);
  }

  resumeSandbox(sandboxId: SandboxId): Promise<void> {
    return this.sandboxes.resumeSandbox(sandboxId);
  }

  /**
   * Renew expiration by setting expiresAt to now + timeoutSeconds.
   */
  async renewSandbox(sandboxId: SandboxId, timeoutSeconds: number): Promise<void> {
    const expiresAt = new Date(Date.now() + timeoutSeconds * 1000).toISOString();
    await this.sandboxes.renewSandboxExpiration(sandboxId, { expiresAt });
  }

  /**
   * Release the HTTP agent resources allocated for this manager instance.
   *
   * Each manager clone owns a scoped `ConnectionConfig` clone.
   *
   * This mirrors the Python SDK's default transport lifecycle.
   */
  async close(): Promise<void> {
    await this.connectionConfig.closeTransport();
  }
}

File: sdks/sandbox/python/src/opensandbox/manager.py
#
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""
Sandbox management interface for administrative operations.

This module provides a centralized interface for managing sandbox instances,
enabling administrative operations and sandbox discovery following the Kotlin SDK pattern.
"""

import logging
from datetime import datetime, timedelta, timezone

from opensandbox.adapters.factory import AdapterFactory
from opensandbox.config import ConnectionConfig
from opensandbox.models.sandboxes import (
    PagedSandboxInfos,
    SandboxFilter,
    SandboxInfo,
    SandboxRenewResponse,
)
from opensandbox.services.sandbox import Sandboxes

logger = logging.getLogger(__name__)


class SandboxManager:
    """
    Sandbox management interface for administrative operations and monitoring sandbox instances.

    This class provides a centralized interface for managing sandbox instances,
    enabling administrative operations and sandbox discovery.
    It focuses on high-level management operations rather than individual sandbox interactions.

    Key Features:

    - **Sandbox Discovery**: List and filter sandbox instances by various criteria
    - **Administrative Operations**: Individual sandbox management operations
    - **Connection Pool Management**: Efficient HTTP client reuse for multiple operations

    Usage Example:

    ```python
    # Create manager
    manager = await SandboxManager.create(connection_config=connection_config)

    # List all running sandboxes
    running_sandboxes = await manager.list_sandbox_infos(
        SandboxFilter(states=["RUNNING"])
    )

    # Individual operations
    sandbox_id = "sandbox-id"
    await manager.get_sandbox_info(sandbox_id)
    await manager.pause_sandbox(sandbox_id)
    await manager.resume_sandbox(sandbox_id)
    await manager.kill_sandbox(sandbox_id)

    # Cleanup
    await manager.close()
    ```

    **Note**: This class is designed for administrative operations.
    For individual sandbox interactions, use the Sandbox class directly.
    """

    def __init__(
        self,
        sandbox_service: Sandboxes,
        connection_config: ConnectionConfig,
    ) -> None:
        """
        Internal constructor for SandboxManager.

        Note: Use SandboxManager.create() instead.

        Args:
            sandbox_service: Service for sandbox operations
            connection_config: Connection configuration (shared transport, headers, timeouts)
        """
        self._sandbox_service = sandbox_service
        self._connection_config = connection_config

    @property
    def connection_config(self) -> ConnectionConfig:
        """Provides access to the connection configuration (including shared transport)."""
        return self._connection_config

    @classmethod
    async def create(
        cls, connection_config: ConnectionConfig | None = None
    ) -> "SandboxManager":
        """
        Creates a SandboxManager instance with the provided configuration.

        Args:
            connection_config: Connection configuration for the manager.
                             If None, default configuration will be used.

        Returns:
            SandboxManager: Configured sandbox manager instance
        """
        config = (connection_config or ConnectionConfig()).with_transport_if_missing()
        factory = AdapterFactory(config)
        sandbox_service = factory.create_sandbox_service()
        return cls(sandbox_service, config)

    async def list_sandbox_infos(self, filter: SandboxFilter) -> PagedSandboxInfos:
        """
        List sandboxes with filtering options.

        Args:
            filter: Filter criteria for sandbox listing

        Returns:
            Paged sandbox information matching the filter criteria

        Raises:
            SandboxException: if the operation fails
        """
        return await self._sandbox_service.list_sandboxes(filter)

    async def get_sandbox_info(self, sandbox_id: str) -> SandboxInfo:
        """
        Get information for a single sandbox by its ID.

        Args:
            sandbox_id: Sandbox ID to retrieve information for

        Returns:
            SandboxInfo for the specified sandbox

        Raises:
            SandboxException: if the operation fails
        """
        logger.debug(f"Getting info for sandbox: {sandbox_id}")
        return await self._sandbox_service.get_sandbox_info(sandbox_id)

    async def kill_sandbox(self, sandbox_id: str) -> None:
        """
        Terminate a single sandbox.

        Args:
            sandbox_id: Sandbox ID to terminate

        Raises:
            SandboxException: if the operation fails
        """
        logger.info(f"Terminating sandbox: {sandbox_id}")
        await self._sandbox_service.kill_sandbox(sandbox_id)
        logger.info(f"Successfully terminated sandbox: {sandbox_id}")

    async def renew_sandbox(self, sandbox_id: str, timeout: timedelta) -> SandboxRenewResponse:
        """
        Renew expiration time for a single sandbox.

        The new expiration time will be set to the current time plus the provided duration.

        Args:
            sandbox_id: Sandbox ID to renew
            timeout: Duration to add to the current time to set the new expiration

        Raises:
            SandboxException: if the operation fails
        """
        # Use timezone-aware UTC datetime to avoid cross-timezone ambiguity.
        new_expiration = datetime.now(timezone.utc) + timeout
        logger.info(f"Renew expiration for sandbox {sandbox_id} to {new_expiration}")
        return await self._sandbox_service.renew_sandbox_expiration(
            sandbox_id, new_expiration
        )

    async def pause_sandbox(self, sandbox_id: str) -> None:
        """
        Pause a single sandbox while preserving its state.

        Args:
            sandbox_id: Sandbox ID to pause

        Raises:
            SandboxException: if the operation fails
        """
        logger.info(f"Pausing sandbox: {sandbox_id}")
        await self._sandbox_service.pause_sandbox(sandbox_id)

    async def resume_sandbox(self, sandbox_id: str) -> None:
        """
        Resume a previously paused sandbox.

        Args:
            sandbox_id: Sandbox ID to resume

        Raises:
            SandboxException: if the operation fails
        """
        logger.info(f"Resuming sandbox: {sandbox_id}")
        await self._sandbox_service.resume_sandbox(sandbox_id)

    async def close(self) -> None:
        """
        Close local resources associated with this sandbox manager.

        This method closes HTTP client resources and other local resources.

        Note: This method logs errors but does not raise exceptions to avoid
        issues in context manager cleanup.
        """
        try:
            # Close transport only when SDK owns it (default transport).
            await self._connection_config.close_transport_if_owned()
        except Exception as e:
            logger.warning(
                f"Error closing resources for sandbox manager: {e}",
                exc_info=True
            )

    async def __aenter__(self) -> "SandboxManager":
        """Async context manager entry."""
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: object,
    ) -> None:
        """Async context manager exit."""
        await self.close()
